### PR TITLE
Add Iconresolver

### PR DESCRIPTION
**Pull request by santonelli**

### CODE REVIEW

File: Products/CMFPlone/browser/iconresolver.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    def __init__(self, context, name):
+        self.context = context
+        self.name = name
+
+    def __call__(self):
+        registry = getUtility(IRegistry)
+        prefix = 'plone.staticresources.icon.'
+        icon = prefix + self.name
+        if icon in registry:
+            return registry[icon]
```
Comment (jensens) [inline Products/CMFPlone/browser/iconresolver.py:27]: If not possible to lookup a LocationError would be nice
```suggestion
            return registry[icon]
        raise LocationError(self.name)
```

File: Products/CMFPlone/browser/iconresolver.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    def __init__(self, context, name):
+        self.context = context
+        self.name = name
+
+    def __call__(self):
+        registry = getUtility(IRegistry)
+        prefix = 'plone.staticresources.icon.'
+        icon = prefix + self.name
+        if icon in registry:
```
Comment (jensens) [inline Products/CMFPlone/browser/iconresolver.py:26]:
```suggestion


PREFIX = 'plone.staticresources.icon.'

@implementer(IPublishTraverse)
class IconBaseResolver:

    def __init__(self, context, name):
        self.context = context
        self.name = name

    def __call__(self):
        registry = getUtility(IRegistry)
        icon = PREFIX + self.name
        if icon in registry:
```

Comment (ale-rt) [reply Products/CMFPlone/browser/iconresolver.py:26]: I would set the `PREFIX` as a class variable:
```
class IconBaseResolver:
     prefix = 'plone.staticresources.icon.'
```
IMO it is easier to override.

Comment (jensens) [reply Products/CMFPlone/browser/iconresolver.py:26]: ok, good point

**Comment by jensens** [comment]
Some basic tests are missing, right?

File: Products/CMFPlone/browser/iconresolver.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                alt_tag = self.name,
+            )            
+        return tag
+
+
+@implementer(ITraversable)
+@adapter(ISiteRoot, Interface)
+class IconBaseTraverser:
+
+    resolver = None
```
Comment (ale-rt) [inline Products/CMFPlone/browser/iconresolver.py:66]: Do we really need another component here? How do you plan to use this one? Before merging I would like to see some documentation draft to understand the big picture.

File: Products/CMFPlone/browser/iconresolver.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        if icon in registry:
+            return registry[icon]
+        raise LocationError(self.name)
+
+
+class IconURLResolver(IconBaseResolver):
+
+    def __call__(self):
+        value = super().__call__()
+        url = getSite().absolute_url() + '/' + value
```
Comment (ale-rt) [inline Products/CMFPlone/browser/iconresolver.py:36]: I think here you want the navigation root URL from plone portal state: https://github.com/plone/plone.app.layout/blob/5f597e41c413c1584a25f01347ead48da52d77c4/plone/app/layout/globals/portal.py#L67
Example: https://github.com/plone/Products.CMFPlone/blob/c3c7ec9ea52fde4d1e868f7df703a80cc8325c0a/Products/CMFPlone/browser/ploneview.py#L156-L160

File: Products/CMFPlone/browser/iconresolver.py (hunk 3 above)
Comment (santonelli) [reply Products/CMFPlone/browser/iconresolver.py:66]: The big picture is described in the issue #3162. This PR is one of two possible solutions. Please consider PR #3194 as prefered.

File: Products/CMFPlone/browser/iconresolver.py (hunk 4 above)
Comment (santonelli) [reply Products/CMFPlone/browser/iconresolver.py:36]: I'm not sure. URL is used to construct the src of e.g. an image-tag. I expcected absolute_url of site is right, when using ++resolveiconurl++some-icon.

File: Products/CMFPlone/browser/iconresolver.py (hunk 3 above)
Comment (ale-rt) [reply Products/CMFPlone/browser/iconresolver.py:66]: Thanks!

**Comment by santonelli** [comment]
Followed by #3194

**Pull request closed without merging**